src/JestProcessManagement/JestProcessManager.ts

```diff
@@ -82,9 +74,21 @@ export class JestProcessManager {
     }
   ): JestProcess {
     if (watch && this.runAllTestsFirstInWatchMode) {
-      return this.handleWatchMode(exitCallback, keepAlive)
+      return this.runAllTestsFirst(exitedJestProcess => {
+        this.removeJestProcessReference(exitedJestProcess)
+        const jestProcessInWatchMode = this.run({
+          watch: true,
+          keepAlive: keepAlive,
+          exitCallback: exitCallback,
+        })
+        exitCallback(exitedJestProcess, jestProcessInWatchMode)
+      })
     } else {
-      return this.handleNonWatchMode(watch, exitCallback, keepAlive)
+      return this.run({
+        watch: watch,
+        keepAlive: keepAlive,
+        exitCallback: exitCallback,
+      })
```

Comment:
Thanks @marcinczenko. That's a lot easier to read.

Reply:
@seanpoulter Sorry for small delay on my side this time. The remaining changes are on the way!

Reply:
Don't be sorry @marcinczenko. It's a big change! After I wrap up some integration tests for `jest-editor-support`, I'll try and write some integration tests for the status bar and output channel.

Reply:
Just added restoring jest editor support events. These are not big changes anymore, bit to write tests and get it tested always requires a bit more time. The last thing to add is to notify the user that the allowed number of Jest process restarts has been reached. I hope to have it done by the end of Monday. Then I think we will be good to go ;).